src/librustc/ty/sty.rs

```diff
@@ -631,16 +629,16 @@ pub enum UpvarSubsts<'tcx> {
 
 impl<'tcx> UpvarSubsts<'tcx> {
     #[inline]
-    pub fn upvar_tys(
-        self,
-        def_id: DefId,
-        tcx: TyCtxt<'tcx>,
-    ) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
-        let upvar_kinds = match self {
-            UpvarSubsts::Closure(substs) => substs.as_closure().split(def_id, tcx).upvar_kinds,
-            UpvarSubsts::Generator(substs) => substs.as_generator().split(def_id, tcx).upvar_kinds,
+    pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
+        let tupled_upvars_ty = match self {
+            UpvarSubsts::Closure(substs) => substs.as_closure().tupled_upvars_ty(),
+            UpvarSubsts::Generator(substs) => substs.as_generator().tupled_upvars_ty(),
+        };
+        let upvars = match tupled_upvars_ty.kind {
+            Tuple(upvars) => upvars,
+            _ => bug!("upvars should be tupled"),
         };
-        upvar_kinds.iter().map(|t| {
+        upvars.iter().map(|t| {
             if let GenericArgKind::Type(ty) = t.unpack() {
```

Comment:
The entire closure body can be `t.expect_ty()`

Reply:
Pre-existing, I wonder what @nikomatsakis and @Zoxc think about it.

Reply:
You can probably just call `tuple_fields` here. Not sure why we use `SubstsRef` for tuples.

Reply:
`SubstsRef` is used for tuples due to the refactoring in https://github.com/rust-lang/rust/issues/42340.